File: Day 2/Advent of Code - Day 2.R
# Load require packages
require(stringr)

# Load the dataset
setwd("~/Data Science/Advent of Code/2018/Day 2")
boxes <- read.csv('boxId.csv', stringsAsFactors = FALSE)

# Part 1: Scan the boxes
alphabet <- c('a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j', 'k', 'l', 'm', 'n', 'o', 'p', 'q', 'r', 's', 't', 'u', 'v', 
              'w', 'x', 'y', 'z')

# Calculate nunmber of boxes that have two or three letters
for(i in alphabet){
  boxes[[i]] <- str_count(boxes$id, i)
}

boxesFlag <- boxes %>% 
  mutate(twoLetter = Reduce('|', lapply(boxes[2:27], '==', 2)),
         threeLetter = Reduce('|', lapply(boxes[2:27], '==', 3)))

# Checksum
boxesFlag %>% filter(twoLetter == TRUE) %>% summarise(n()) * boxesFlag %>% filter(threeLetter == TRUE) %>% summarise(n())

# Part 2: finding the right letters
# Iterate over the rows and each character in the string to find the duplicate password
pass <- data.frame()
for(j in 1:nrow(boxes)){
  for(i in 1:26){
    front <- substr(boxes[j, 1], 1, i-1)
    mid <- str_replace(substr(boxes[j, 1], i, i), '[a-z]', '_')
    back <- substr(boxes[j, 1], i+1, 26)
    newStr <- paste0(front, mid, back)
    newStr <- data.frame('result' = newStr)
    pass <- pass %>% bind_rows(newStr)
  }
}

# Print the result
pass$result[duplicated(pass$result)]

File: Day 2/Advent of Code - Day 2.py
# Advent of Code 2018 - Day 2
import os 
import pandas as pd
import itertools
import collections

# Load the dataset
os.chdir('C:/Users/malon/Documents/Data Science/Advent of Code/2018/Adventofcode2018/Day 2')
boxes = pd.read_csv('boxId.csv')

# Part 1
def day2_1(input: str):
    ntimes = {2:0, 3:0}
    for i in input:
        letterNtimes = {}
        collections._count_elements(letterNtimes, i)
        nTimesSet = set(list(letterNtimes.values()))
        nTimesSet.discard(1)
        collections._count_elements(ntimes, nTimesSet)
    checksum = ntimes[2] * ntimes[3]
    return checksum

day2_1(boxes['id'])

# Part 2
def day2_2(input: str):

    def differentByOneLetter(s1, s2):
        alreadyDifferent = False
        for c1, c2 in zip(s1, s2):
            if c1 != c2:
                if alreadyDifferent:
                    return False
                else:
                    alreadyDifferent = True

        return True

    def commonLetters(s1, s2):
        for index, (c1, c2) in enumerate(zip(s1, s2)):
            if c1 != c2:
                return s1[:index] + s1[index+1:]

    for s1, s2 in itertools.combinations(input, 2):
        areDifferent = differentByOneLetter(s1, s2)
        if areDifferent:
            commonLetters = commonLetters(s1, s2)
            return commonLetters

day2_2(boxes['id'])